Source/SwiftLintFramework/Rules/Lint/TypesafeArrayInitRule.swift

```diff
@@ -0,0 +1,131 @@
+            return []
+        }
+        let index = buildIndex(for: filePath, using: compilerArguments)
+        return index.traverseEntitiesDepthFirst { substructure -> [StyleViolation]? in
+            guard substructure.kind == "source.lang.swift.ref.function.method.instance",
+                  let line = substructure.line, let column = substructure.column, let filePath = file.path else {
+                return nil
+            }
+            let offset = file.stringView.byteOffset(forLine: Int(line), column: Int(column))
+            let cursorInfoRequest = Request.cursorInfo(file: filePath, offset: offset, arguments: compilerArguments)
+            guard let cursorInfo = try? cursorInfoRequest.sendIfNotDisabled(),
+                  let isSystem = cursorInfo["key.is_system"], isSystem.isEqualTo(true),
+                  let name = cursorInfo["key.name"], name.isEqualTo("map(_:)"),
+                  let typeName = cursorInfo["key.typename"] as? String,
+                  Self.mapTypePattern.numberOfMatches(in: typeName, range: NSRange(for: typeName)) == 1,
+                  let dict = pickSubstructure(from: file.structureDictionary, at: offset) else {
+                return nil
+            }
+            return Self.parentRule.validate(file: file, kind: .call, dictionary: dict)
+        }.flatMap { $0 }
+    }
+
+    private func buildIndex(for filePath: String, using compilerArguments: [String]) -> SourceKittenDictionary {
+        do {
+            return SourceKittenDictionary(
+                try Request.index(file: filePath, arguments: compilerArguments).sendIfNotDisabled()
+            )
+        } catch {
+            queuedPrintError("""
+                Indexing of file '\(filePath)' in the context of the \(Self.description.identifier) rule failed.
+                """)
+        }
+        return SourceKittenDictionary([:])
+    }
+
+    private func pickSubstructure(from: SourceKittenDictionary, at mapOffset: ByteCount) -> SourceKittenDictionary? {
+        let substructures = from.traverseBreadthFirst { substructure -> [SourceKittenDictionary]? in
+            guard substructure.expressionKind == .call,
+                  let name = substructure.name, name.hasSuffix(".map"),
+                  let nameOffset = substructure.nameOffset,
+                  let nameLength = substructure.nameLength,
+                  mapOffset + ByteCount("map".count) == nameOffset + nameLength else {
+                return nil
+            }
+            return [substructure]
+        }
+        return substructures.count == 1 ? substructures.first : nil
+    }
+}
+
+private extension NSRange {
+    init(for string: String) {
+        self.init(location: 0, length: string.utf16.count)
+    }
+}
+
+private extension StringView {
+    func byteOffset(forLine line: Int, column: Int) -> ByteCount {
+        guard line > 0 else { return ByteCount(column - 1) }
+        return lines[line - 1].byteRange.location + ByteCount(column - 1)
+    }
+}
```

Comment:
This is now duplicated 4 times, can you make a new `Source/SwiftLintFramework/Extensions/StringView+SwiftLint.swift` file with this?

Reply:
It's already on my todo list. I guess it is fine if I refactor that in a separate change afterwards to not mix things up.

Reply:
#3917